### PR TITLE
os.sleep: change type to: `milsecs: Natural`

### PR DESCRIPTION
**Pull request by juancarlospaco**
- Changed `os.sleep` argument from `int` to `Natural`, because negative values make no sense.
- You can not sleep negative time, is buggy and breaks laws of thermodynamics or something.

### CODE REVIEW

**Review by dom96 — COMMENTED** [review]
`Natural` is a range (https://nim-lang.org/docs/system.html#Natural), and I know that at least Status considers them to be unsafe so this may not be wise.

**Comment by juancarlospaco** [comment]
We should Fix ranges then... :shrug: 
Everything buggy is unsafe and unstable thats why is buggy. At least a range should not be too hard to test (?).

**Comment by timotheecour** [comment]
> Natural is a range (nim-lang.org/docs/system.html#Natural), and I know that at least Status considers them to be unsafe so this may not be wise.

@dom96 can you please provide a link? otherwise this is too vague. What do you/they mean by unsafe?
see also https://github.com/nim-lang/RFCs/issues/270

> in nim-lang/Nim@8ee0771#commitcomment-38239130 araq argued "return types must not be Natural" (this came up recently here: nim-lang/fusion#18 (comment)), and instead suggested using {.ensures: cond.} for that.

but this concerns return types, not parameters.

File: tests/osproc/texitsignal.nim [hunk 1]
```diff
@@ -36,4 +36,4 @@ if paramCount() == 0:
     echo p.waitForExit() == 128 + SIGKILL
 
 else:
-  sleep(5000)  # should get killed before this
+  os.sleep(5000)  # should get killed before this
```
Comment (timotheecour) [inline tests/osproc/texitsignal.nim:39]: why was this change needed?

Comment (juancarlospaco) [reply tests/osproc/texitsignal.nim:39]: Test was using the wrong `sleep`, because both `os` and `posix` have it, fully qualified is as explicit as it gets,
whatsoever the test is not about sleeps https://github.com/nim-lang/Nim/pull/17149/commits/4464d04dc8aceed03f674df30ba114a24a82c1b4

**Comment by juancarlospaco** [comment]
:bug:  https://play.nim-lang.org/#ix=2QrL

**Comment by timotheecour** [comment]
@juancarlospaco please make the title a lot less generic, for eg: `os.sleep: change type to Natural` (generally applicable)

EDIT i did it for you

File: tests/osproc/texitsignal.nim (hunk 1 above)
Comment (Araq) [reply tests/osproc/texitsignal.nim:39]: If the change breaks our very own test suite, it'll probably also break somebody else's code. Sorry, but this has to be rejected because of this.

**Comment by Araq** [comment]
Causes subtle code breakage, unfortunately.

**Comment by timotheecour** [comment]
@Araq  indeed: this breaks the following code because `Natural` (a range) matches less well than cint for `sleep(1)`.

```nim
import os
import posix

proc main() =
  echo "ok1"
  sleep(3)
  echo "ok2"
  discard posix.sleep(3)
  echo "ok3"
  os.sleep(3)
  echo "ok4"
main()
```

But the fact that we have 2 procs named the same (os.sleep, posix.sleep) but with different semantics (milliseconds vs seconds) is IMO error prone (as demonsrated in this PR). At least here we're saved by the fact that one proc returns void and not the other but depending on situations that wouldn't give CT error, but would silently call the wrong proc with the wrong unit.


## proposal
introduce `proc sleep(t: TimeInterval)` in std/times with following benefits:
* no implicit conversion possible
* self-documenting units at call site:
```nim
sleep(1.seconds)
sleep(1.milliseconds)
sleep(100.milliseconds + 1.seconds)
```

note that std/times is a heavy dependency, but pending nim-lang/RFCs#55 users who need `times.sleep` could import a ligthter-weight module (that times would re-export)

Furthermore, with https://github.com/nim-lang/RFCs/issues/270 we'd also be able to write:
```nim
# os.nim:
proc sleep(t: int) {.ensures t>=0.}

# times.nim
proc sleep(t: TimeInterval) {.ensures t>=0.}
```

now we'd have a sane API.

File: tests/osproc/texitsignal.nim (hunk 1 above)
Comment (juancarlospaco) [reply tests/osproc/texitsignal.nim:39]: Documentation explicitly says to use fully qualified names when code is ambiguous. :shrug:

**Comment by juancarlospaco** [comment]
Why not just use `uint` ?.

- No implicit conversion possible
- No need for units
- No need for ensures
- No times dependency
- No need for --checks:on
- No negative even with -d:danger

If wrong, please provide a repro code to make a Negative `uint`.

**Comment by juancarlospaco** [comment]
Still no repro code or link for `Natural` being broken on non-Danger mode.

**Comment by timotheecour** [comment]
> Why not just use uint ?.
> If wrong, please provide a repro code to make a Negative uint.

`uint` can't work for this purpose:
```nim
proc sleep2(a: uint) = echo a
var t = -3
sleep2(t.uint) # 18446744073709551613
```

> Still no repro code or link for Natural being broken on non-Danger mode.

Natural is a range, changing a proc param from int to range changes how it matches (affects sigmatch).

I'm proposing `{.requires.}` (https://github.com/nim-lang/RFCs/issues/270) as a replacement, which doesn't affect the type system/sigmatch and is more flexible.

the other issue I raised regarding units is a separate issue and both can be addressed independently.

**Comment by juancarlospaco** [comment]
That confirms what I am saying

> ```nim
> proc sleep2(a: uint) = echo a
> var t = -3
> sleep2(t.uint) # 18446744073709551613
> ```

**Inside the function**
- Theres no Negative.
- Theres no Bug.

You are producing a bug *before* entering the proc :shrug:

**Comment by timotheecour** [comment]
This doesn't make sense. API's should be easy to use, hard to misused; the fact that you allow valid code (no warnings, no errors) to produce a bug means that API is bad.

**Comment by juancarlospaco** [comment]
Theres no bug because `18446744073709551613` is the correct value for `-3` **on the `uint` type**
otherwise that means that integers and basic math is completely broken on the language.

But I agree, and ultimately is Ok, because the relevant thing is that `Natural` should be fixed if it is really broken.

**Comment by konsumlamm** [comment]
> ```nim
> proc sleep2(a: uint) = echo a
> var t = -3
> sleep2(t.uint) # 18446744073709551613
> ```

That kind of argument can be used against pretty much any use of `uint` though... So it seems like either `uint` shouldn't exist, or `t.uint` should throw an exception if `t` is negative.

**Comment by timotheecour** [comment]
the main use cases for unsigned int types are:
* interfacing with FFI
* types for which wraparound behavior is desirable
* low level bit manipulation code
* if you need to accomodate numbers > int.high

One important place where unsigned would make more sense than signed is `hashes.Hash`, since hashes are almost always operated on by bit manipulation, not comparison, and negative hashes don't make sense (and wrap around behavior is desirable).


## note
this is also consistent with the C++ google style guide, see https://google.github.io/styleguide/cppguide.html#Integer_Types ; and I agree with that section:

> You should not use the unsigned integer types such as uint32_t, unless there is a valid reason such as representing a bit pattern rather than a number, or you need defined overflow modulo 2^N. In particular, do not use unsigned types to say a number will never be negative. Instead, use assertions for this.

> Unsigned integers are good for representing bitfields and modular arithmetic. Because of historical accident, the C++ standard also uses unsigned integers to represent the size of containers - many members of the standards body believe this to be a mistake, but it is effectively impossible to fix at this point. The fact that unsigned arithmetic doesn't model the behavior of a simple integer, but is instead defined by the standard to model modular arithmetic (wrapping around on overflow/underflow), means that a significant class of bugs cannot be diagnosed by the compiler. In other cases, the defined behavior impedes optimization.

## TODO
- [ ] we should document the fact that most API's should prefer signed instead of unsigned types, explain the pitfalls and when to use unsigned types; a good place could be in contributing.rst, cross referenced from the `system.uint` declaration

**Comment by Araq** [comment]
> and I agree with that section:...

Correct and it's why we use `int` for `len`. Which is also what C# does btw and a full class of bugs is pretty obscure in C# because of this (`for (int i = 0; i < n.length - 3; ++i)`) whereas common in C(++). If that's not enough of a justification, I don't know what is.
